File: Source/EmbeddedSw/Commands/CommandBase.cpp
/*******************************************************************
@copyright COPYRIGHT AND PROPRIETARY RIGHTS NOTICES:

Copyright (C) 2021, an unpublished work by Syncroness, Inc.
All rights reserved.

This material contains the valuable properties and trade secrets of
Syncroness of Westminster, CO, United States of America
embodying substantial creative efforts and confidential information,
ideas and expressions, no part of which may be reproduced or
transmitted in any form or by any means, electronic, mechanical, or
otherwise, including photocopying and recording or in connection
with any information storage or retrieval system, without the prior
written permission of Syncroness.
****************************************************************** */


/**
 * Implementation of CommandBase functions
 * Note:  Pure virtual functions can cause the compiler/linker to pull in
 * an excess amount of code.  So, pure virtual functions should be implemented to
 * fail at run time instead (not ideal, but the least bad option).
 */

#include "CommandBase.hpp"
#include "Logging.hpp"

// Initialize the rolling sequence number that will make each object instantiation have a quasi unique id
CommandBase::commandSequenceNumber_t CommandBase::m_rollingCommandSequenceNumber = 0;


bool CommandBase::execute(void* p_childCommand)
{
	LOG_FATAL(Logging::LogModuleIdCefInfrastructure, "Base class execute() called, supposed to be implemented in derived class");
	return true;
}

errorCode_t CommandBase::importFromCefCommand(void* p_cefCommand)
{
	LOG_FATAL(Logging::LogModuleIdCefInfrastructure, "Base class importFromCefCommand() called, supposed to be implemented in derived class");
	return errorCode_LogFatalReturn;
}

errorCode_t CommandBase::importFromCefCommandBase(cefCommandHeader_t* p_cefCommandHeader, uint32_t numBytesInCefRequestCommand)
{
	if (p_cefCommandHeader == nullptr)
	{
		LOG_FATAL(Logging::LogModuleIdCefInfrastructure, "Base class importFromCefCommand called with nullptr");
		return errorCode_PointerIsNullptr;
	}

	// Sanity Check that have the right opcode
	if (p_cefCommandHeader->m_commandOpCode != m_commandOpCode)
	{
		LOG_FATAL(Logging::LogModuleIdCefInfrastructure, "Base class importFromCefCommand cefOpCode(0x{:X}) != expected OpCode (0x{:X})", p_cefCommandHeader->m_commandOpCode, m_commandOpCode);
		return errorCode_CmdBaseImportCefCommandOpCodeDoesNotMatchCommand;
	}

	// Sanity Check that python utilities and embedded software are in synch
	if (p_cefCommandHeader->m_commandNumBytes != numBytesInCefRequestCommand)
	{
		LOG_FATAL(Logging::LogModuleIdCefInfrastructure, "Base class importFromCefCommand m_commandNumBytes(0x{:X}) != expected numBytesInCefRequestCommand (0x{:X})", m_commandNumBytes, numBytesInCefRequestCommand);
		return errorCode_CmdBaseImportCefCommandNumBytesInCefRequestDoesNotMatch;
	}

	return errorCode_OK;
}

errorCode_t CommandBase::exportToCefCommand(void* p_cefCommand)
{
	LOG_FATAL(Logging::LogModuleIdCefInfrastructure, "Base class exportToCefCommand() called, supposed to be implemented in derived class");
	return errorCode_LogFatalReturn;
}

errorCode_t CommandBase::exportToCefCommandBase(cefCommandHeader_t* p_cefCommandHeader, uint32_t numBytesInCefResponseCommand)
{
	if (p_cefCommandHeader == nullptr)
	{
		LOG_FATAL(Logging::LogModuleIdCefInfrastructure, "Base class importFromCefCommand called with nullptr");
		return errorCode_PointerIsNullptr;
	}

	p_cefCommandHeader->m_commandOpCode = m_commandOpCode;

	// Add number of bytes in CefResponseCommand to Python Utilities can do a sanity check
	p_cefCommandHeader->m_commandNumBytes = numBytesInCefResponseCommand;

	p_cefCommandHeader->m_commandErrorCode = m_commandErrorCode;

	return errorCode_OK;
}

File: Source/Python/Commands/CommandBase.py
# ##################################################################
#\copyright COPYRIGHT AND PROPRIETARY RIGHTS NOTICES:
#
#Copyright (C) 2021, an unpublished work by Syncroness, Inc.
#All rights reserved.
#
#This material contains the valuable properties and trade secrets of
#Syncroness of Westminster, CO, United States of America
#embodying substantial creative efforts and confidential information,
#ideas and expressions, no part of which may be reproduced or
#transmitted in any form or by any means, electronic, mechanical, or
#otherwise, including photocopying and recording or in connection
#with any information storage or retrieval system, without the prior
#written permission of Syncroness.
################################################################## #


import sys
from os.path import dirname, abspath
from abc import ABC, abstractmethod

sys.path.append(dirname(dirname(abspath(__file__))))
from Shared import cefContract


class CommandBase(ABC):
    """
    Partially abstract base class for CEF commands. All commands have a header and body.
    The body can be either a request (originating from the Python Utility side) or a response
    to that request originating from the embedded target. Both use the same struct for header.
    """
    def __init__(self):
        self.header = cefContract.cefCommandHeader()
        self.request = None
        self.expectedResponse = None

    @abstractmethod
    def buildCommand(self):
        """
        Populate the command header and request body according to the CEF contract. This should make
        use of any and all command-specific expected/test values for both requests and responses.
        """
        pass

    @abstractmethod
    def validateResponseBody(self):
        """
        Command-specific field value checking
        """
        pass

    @abstractmethod
    def expectedResponseType(self):
        """
        Command-specific struct to validate against upon response
        """
        pass

    def validateResponseHeader(self, responseHeader: cefContract.cefCommandHeader):
        """
        Common header field value checking. The response sequence number must match what was sent out, 
        the command must have executed on-target without error, and the response OpCode must match what was sent.
        @param responseHeader: the command header obeying the structure defined in cefContract
        @return boolean: False if any check fails, else True
        """
        if responseHeader.m_commandSequenceNumber != self.header.m_commandSequenceNumber:
            print("Response Sequence Number mis-match - received: {}, sent: {}".format(responseHeader.m_commandSequenceNumber, self.header.m_commandSequenceNumber))
            return False
        if responseHeader.m_commandErrorCode != cefContract.errorCode.errorCode_OK.value:
            print("Response Error Code: {}".format(responseHeader.m_commandErrorCode))
            return False
        if responseHeader.m_commandOpCode != self.header.m_commandOpCode:
            print("Response OpCode mis-match - received: {}, sent: {}".format(responseHeader.m_commandOpCode, self.header.m_commandOpCode))
            return False
        return True

    def __len__(self):
        return len(bytes(self.request))

    def payload(self):
        return bytes(self.request)

    def setSequenceNumber(self, sequenceNumber):
        self.header.m_commandSequenceNumber = sequenceNumber

    def expectedResponseLength(self):
        return len(bytes(self.expectedResponse))
